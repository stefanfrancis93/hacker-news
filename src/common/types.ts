export interface ApiResponse<T> {
  data: T;
  status: number;
  error?: Error;
}

export interface Error {
  message?: string;
}

export enum Type {
  STORY = "story",
}

export interface Story {
  by: string;
  descendants: number;
  id: number;
  kids: number[];
  score: number;
  time: number;
  title: string;
  text?: string;
  type: Type;
  url?: string;
}

export interface Pagination {
  page: number;
  limit: number;
}
