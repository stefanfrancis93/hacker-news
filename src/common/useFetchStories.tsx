import { useState, useEffect, useCallback } from "react";
import { ApiResponse, Pagination, Story } from "@/common/types";

const useFetchStories = (page: number) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [stories, setStories] = useState<Story[]>([]);
  const [hasMore, setHasMore] = useState(false);

  useEffect(() => {
    if (page === 1) {
      setStories([]);
    }
  }, [page]);

  const fetchData = useCallback(
    async (url: URL) => {
      const response = await fetch(url);
      const responseJson: ApiResponse<{
        stories: Story[];
        pagination: Pagination;
      }> = await response.json();
      console.log(page);

      if (page === 1) {
        setStories(responseJson.data.stories);
      } else {
        setStories((prev: Story[]) => [...prev, ...responseJson.data.stories]);
      }
      setHasMore(responseJson.data.stories.length > 0);
      setLoading(false);
    },
    [page]
  );

  useEffect(() => {
    try {
      setLoading(true);
      setError(false);
      let url = new URL("http://localhost:3000/api/stories/top");
      url.searchParams.set("page", page.toString());

      fetchData(url);

    } catch (err: unknown) {
      setError(!!err);
    }
  }, [fetchData, page]);

  return { loading, error, hasMore, stories };
};

export default useFetchStories;
