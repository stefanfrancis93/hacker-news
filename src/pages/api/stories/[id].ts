import { Story, ApiResponse } from "@/common/types";
import type { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
  _req: NextApiRequest,
  res: NextApiResponse<ApiResponse<Story | null>>
) {
  const { query, method } = _req;
  const { id } = query;

  // Only allow GET requests
  if (method !== "GET") {
    res.status(405).json({
      error: { message: "Method not allowed" },
      status: 405,
      data: null,
    });
    return;
  }

  // id must be present
  if (!id) {
    res.status(400).json({
      error: { message: "Missing id in the query" },
      status: 400,
      data: null,
    });
    return;
  }

  try {
    const { HACKER_NEWS_API_BASE_URL } = process.env;

    const API_URL = `${HACKER_NEWS_API_BASE_URL}/item/${id}.json?print=pretty`;
    const response = await fetch(API_URL);
    const data: Story | null = await response.json();

    if (data) {
      res.status(200).json({ data, status: 200 });
    } else {
      res.status(404).json({
        error: { message: `Story with id: ${id} not found` },
        status: 404,
        data: null,
      });
    }
  } catch (e) {
    res.status(500).json({
      error: { message: "Failed to load data" },
      status: 500,
      data: null,
    });
  }
}
