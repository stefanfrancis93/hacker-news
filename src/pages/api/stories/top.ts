import { ApiResponse, Pagination, Story } from "@/common/types";
import type { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
  _req: NextApiRequest,
  res: NextApiResponse<
    ApiResponse<{
      stories: Story[];
      pagination: Pagination;
    } | null>
  >
) {
  const { query, method } = _req;

  // Only allow GET requests
  if (method !== "GET") {
    res.status(405).json({
      error: { message: "Method not allowed" },
      status: 405,
      data: null,
    });
    return;
  }

  try {
    const page = parseInt(query.page as string, 10) || 0;
    const limit = parseInt(query.limit as string, 10) | 10;
    const startIndex = page * limit;
    const endIndex = (page + 1) * limit;

    const { BASE_URL, HACKER_NEWS_API_BASE_URL } = process.env;

    const API_URL = `${HACKER_NEWS_API_BASE_URL}/topstories.json?print=pretty`;
    const response = await fetch(API_URL);
    const responseJson: number[] = await response.json();
    let ids = responseJson.slice(startIndex, endIndex);
    const stories = await Promise.all(
      ids.map(async (id: number) => {
        const ITEM_URL = `${BASE_URL}/api/stories/${id}`;
        const itemResponse = await fetch(ITEM_URL);
        const itemResponseJson: ApiResponse<Story | null> =
          await itemResponse.json();
        return itemResponseJson.data;
      })
    );

    res.status(200).json({
      data: {
        stories: stories.filter((story) => !!story) as Story[],
        pagination: {
          page,
          limit,
        },
      },
      status: 200,
    });
  } catch (e) {
    console.log(e);
    res.status(500).json({
      error: { message: "Failed to load data" },
      status: 500,
      data: null,
    });
  }
}
