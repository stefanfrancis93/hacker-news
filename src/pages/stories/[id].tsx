import { useEffect, useState } from "react";
import { ApiResponse, Story } from "@/common/types";
import { useRouter } from "next/router";

const StoryDetails = () => {
  const router = useRouter();
  const { id } = router.query;
  const [story, setStory] = useState<Story | null>(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    (async () => {
      if (!id) return; // Skips API call if id is not present
      setLoading(true);
      const response = await fetch(`http://localhost:3000/api/stories/${id}`);
      const responseJson: ApiResponse<Story> = await response.json();
      setStory(responseJson.data);
      setLoading(false);
    })();
  }, [id]);

  if (loading) {
    return <>Loading...</>;
  }

  return <>{story?.title}</>;
};

export default StoryDetails;
