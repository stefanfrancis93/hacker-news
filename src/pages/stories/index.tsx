import { useCallback, useRef, useState } from "react";
import Link from "next/link";
import styles from "../../styles/StoryList.module.css";
import useFetchStories from "@/common/useFetchStories";
import Image from "next/image";

const Stories = () => {
  const [page, setPage] = useState(1);
  const { loading, error, hasMore, stories } = useFetchStories(page);

  const observer = useRef<IntersectionObserver | null>();
  const lastStoryElRef = useCallback(
    (node: HTMLLIElement | null) => {
      if (loading) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting && hasMore) {
          setPage((prev) => prev + 1);
        }
      });
      if (node) observer.current.observe(node);
    },
    [loading, hasMore]
  );

  const getLinkElement = useCallback(
    ({
      title,
      text,
      time,
      url,
    }: {
      title: string;
      text?: string;
      time: number;
      url: string;
    }) => (
      <Link href={url} target="_blank" className={styles.link}>
        <div className={styles.titleContainer}>
          <h2 className={styles.storyTitle}>{title}</h2>
          <span>{new Date(time).toLocaleDateString()}</span>
        </div>
        {text && <p className={styles.storyContent}>{text}</p>}
      </Link>
    ),
    []
  );

  return (
    <main className={styles.main}>
      <h1 data-testid="heading">Hacker News | Top Stories</h1>
      <ul data-testid="list-container" className={styles.list}>
        {stories.map(({ id, title, text, time, url = "" }, index) => {
          const isLastElement = stories.length === index + 1;
          if (isLastElement) {
            return (
              <li
                key={id}
                className={styles.listItem}
                data-testid="story-item"
                ref={lastStoryElRef}
              >
                {getLinkElement({ title, text, time, url })}
              </li>
            );
          } else {
            return (
              <li key={id} className={styles.listItem} data-testid="story-item">
                {getLinkElement({ title, text, time, url })}
              </li>
            );
          }
        })}
      </ul>
      {loading && (
        <div
          data-testid="spinner-container"
          className={styles.spinnerContainer}
        >
          Loading
          <span className={styles.spinner}>
            <Image src="/spinner.gif" alt="spinner" width={30} height={30} />
          </span>
        </div>
      )}
      {error && <div>Error...</div>}
    </main>
  );
};

export default Stories;
