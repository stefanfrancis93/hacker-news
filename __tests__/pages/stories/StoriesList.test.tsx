import { render } from "@testing-library/react";
import StoriesList from "@/pages/stories";
import "@testing-library/jest-dom";
import useFetchStories from "../../../src/common/useFetchStories";
import { firstTen, secondTen } from "./stories.json";
import { Story } from "@/common/types";

jest.mock("../../../src/common/useFetchStories");

const mockUseFetchStories = useFetchStories as jest.MockedFunction<
  typeof useFetchStories
>;
mockUseFetchStories.mockReturnValue({
  loading: true,
  error: false,
  hasMore: false,
  stories: [],
});

beforeEach(() => {
  // IntersectionObserver isn't available in test environment
  const mockIntersectionObserver = jest.fn();
  mockIntersectionObserver.mockReturnValue({
    observe: jest.fn,
    unobserve: jest.fn,
    disconnect: jest.fn,
  });
  window.IntersectionObserver = mockIntersectionObserver;
});

describe("Stories", () => {
  it("renders the page unchanged", () => {
    const { container } = render(<StoriesList />);
    expect(container).toMatchSnapshot();
  });

  it("renders stories list container and spinner initially", () => {
    mockUseFetchStories.mockReturnValue({
      loading: true,
      error: false,
      hasMore: false,
      stories: [],
    });
    const { queryByTestId } = render(<StoriesList />);
    const listContainer = queryByTestId("list-container");
    expect(listContainer).toBeInTheDocument();
    const spinnerContainer = queryByTestId("spinner-container");
    expect(spinnerContainer).toBeInTheDocument();
  });

  it("renders 10 stories after loading", () => {
    mockUseFetchStories.mockReturnValue({
      loading: false,
      error: false,
      hasMore: false,
      stories: firstTen as Story[],
    });
    const { queryByTestId, getAllByTestId } = render(<StoriesList />);
    const storyItems = getAllByTestId("story-item");
    expect(storyItems.length).toEqual(10);
    const spinnerContainer = queryByTestId("spinner-container");
    expect(spinnerContainer).not.toBeInTheDocument();
  });

  it("renders loading when user reaches the last story", () => {
    mockUseFetchStories.mockReturnValue({
      loading: true,
      error: false,
      hasMore: false,
      stories: firstTen as Story[],
    });
    const { queryByTestId, getAllByTestId } = render(<StoriesList />);
    const storyItems = getAllByTestId("story-item");
    expect(storyItems.length).toEqual(10);
    const spinnerContainer = queryByTestId("spinner-container");
    expect(spinnerContainer).toBeInTheDocument();
  });

  it("renders next 10 stories when user reaches the last story", () => {
    mockUseFetchStories.mockReturnValue({
      loading: false,
      error: false,
      hasMore: false,
      stories: [...firstTen, ...secondTen] as Story[],
    });
    const { queryByTestId, getAllByTestId } = render(<StoriesList />);
    const storyItems = getAllByTestId("story-item");
    expect(storyItems.length).toEqual(20);
    const spinnerContainer = queryByTestId("spinner-container");
    expect(spinnerContainer).not.toBeInTheDocument();
  });
});
