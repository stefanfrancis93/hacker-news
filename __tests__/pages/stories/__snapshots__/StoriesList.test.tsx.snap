// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stories renders the page unchanged 1`] = `
<div>
  <main
    class="main"
  >
    <h1
      data-testid="heading"
    >
      Hacker News | Top Stories
    </h1>
    <ul
      class="list"
      data-testid="list-container"
    />
    <div
      class="spinnerContainer"
      data-testid="spinner-container"
    >
      Loading
      <span
        class="spinner"
      >
        <img
          alt="spinner"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/_next/image?url=%2Fspinner.gif&w=64&q=75"
          srcset="/_next/image?url=%2Fspinner.gif&w=32&q=75 1x, /_next/image?url=%2Fspinner.gif&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
      </span>
    </div>
  </main>
</div>
`;
